### PR TITLE
Fix config_content.as_dict()

### LINKED ISSUE #4148: Improve config_content.as_dict()
Handle the creation of the config_content dictionary in a better way so that we avoid having to do the following:

```python 
  max_runtime = config_dict.get(ConfigKeys.MAX_RUNTIME, 0)
  if isinstance(max_runtime, list):
      max_runtime = max_runtime[-1]
```
 Currently, any consumer of the config_content dict needs to account for this behaviour if the config KEY is defined multiple times in the file and we are only interested in the final value.

### PR DESCRIPTION
**Pull request by DanSava**
**Issue**
Resolves #4148

**Approach**
Add a list of single-value config keys for each of the keys in the list config_content.as_dict() will return a single value for the resulting dictionary item.

## Pre review checklist

- [x] Added appropriate release note label
- [x] PR title captures the intent of the changes, and is fitting for release notes.
- [x] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).

Adding labels helps the maintainers when writing release notes. This is the [list of release note labels](https://github.com/equinor/ert/labels?q=release-notes).

### CODE REVIEW

File: src/ert/_c_wrappers/config/config_content.py [hunk 1]
```diff
@@ -7,11 +7,52 @@
...
 
 
+SINGLE_VALUE_KEYS = [
+    ConfigKeys.ANALYSIS_SELECT,
+    ConfigKeys.CONFIG_DIRECTORY,
+    ConfigKeys.DATAROOT,
+    ConfigKeys.DATA_FILE,
+    ConfigKeys.ENSPATH,
+    ConfigKeys.RERUN_KEY,
+    ConfigKeys.ALPHA_KEY,
```
Comment (andreas-el) [inline src/ert/_c_wrappers/config/config_content.py:23]: I have a strong feeling this list was sorted at some point. Suggest to redo that?

Comment (DanSava) [reply src/ert/_c_wrappers/config/config_content.py:23]: Good point, it was never sorted but it should be sorted.

File: tests/unit_tests/c_wrappers/res/config/test_config_content.py [hunk 2]
```diff
@@ -2,12 +2,51 @@
...
 from ert._clib.config_keywords import init_user_config_parser
 
+SINGLE_VALUE_KEY_TYPES = {
+    ConfigKeys.ANALYSIS_SELECT: "str",
+    ConfigKeys.CONFIG_DIRECTORY: "path",
+    ConfigKeys.DATAROOT: "file_path",
+    ConfigKeys.DATA_FILE: "file_path",
+    ConfigKeys.ENSPATH: "path",
+    ConfigKeys.RERUN_KEY: "bool",
+    ConfigKeys.ALPHA_KEY: "int",
```
Comment (andreas-el) [inline tests/unit_tests/c_wrappers/res/config/test_config_content.py:18]: Sorting again

**Review by andreas-el — CHANGES_REQUESTED** [review]

File: src/ert/_c_wrappers/config/config_content.py [hunk 3]
```diff
@@ -14,42 +14,42 @@
...
-    ConfigKeys.NUM_REALIZATIONS,
     ConfigKeys.NUM_CPU,
+    ConfigKeys.NUM_REALIZATIONS,
     ConfigKeys.OBS_CONFIG,
+    ConfigKeys.QUEUE_SYSTEM,
+    ConfigKeys.RANDOM_SEED,
+    ConfigKeys.RERUN_START_KEY,
+    ConfigKeys.RUNPATH,
+    ConfigKeys.RERUN_KEY,
+    ConfigKeys.REFCASE,
```
Comment (andreas-el) [inline src/ert/_c_wrappers/config/config_content.py:47]:
```
  ConfigKeys.RANDOM_SEED,
    ConfigKeys.REFCASE,
    ConfigKeys.RERUN_KEY,
    ConfigKeys.RERUN_START_KEY,
    ConfigKeys.RUNPATH,
    ConfigKeys.RUNPATH_FILE,
```

File: src/ert/_c_wrappers/enkf/analysis_config.py [hunk 4]
```diff
@@ -112,7 +109,7 @@ def from_dict(cls, config_dict) -> "AnalysisConfig":
             std_cutoff=config_dict.get(ConfigKeys.STD_CUTOFF_KEY, 1e-6),
             stop_long_running=config_dict.get(ConfigKeys.STOP_LONG_RUNNING, False),
             global_std_scaling=config_dict.get(ConfigKeys.GLOBAL_STD_SCALING, 1.0),
-            max_runtime=max_runtime,
+            max_runtime=config_dict.get(ConfigKeys.MAX_RUNTIME, 0),
```
Comment (valentin-krasontovitsch) [inline src/ert/_c_wrappers/enkf/analysis_config.py:112]: this change (together with the deleted lines 104-106 above) to how `MAX_RUNTIME` is picked from a config dict seems unrelated to my eyes  - would you mind shortly explaining this change? is it just tidying up, because we do not expect `MAX_RUNTIME` to be a list in the config dict?

File: src/ert/_c_wrappers/config/config_content.py [hunk 5]
```diff
@@ -14,42 +14,42 @@
...
+    ConfigKeys.RUNPATH_FILE,
     ConfigKeys.STOP_LONG_RUNNING,
```
Comment (andreas-el) [inline src/ert/_c_wrappers/config/config_content.py:49]:
```
    ConfigKeys.STD_CUTOFF_KEY,
    ConfigKeys.STOP_LONG_RUNNING,
```

File: tests/unit_tests/c_wrappers/res/config/test_config_content.py [hunk 6]
```diff
@@ -74,9 +112,35 @@ def test_config_content_as_dict(tmpdir):
...
+            "bool": "TRUE",
+            "float": 4.2,
+            "int": 2,
+        }
+
+        with open("config.file", "w") as fileH:
+            for key in SINGLE_VALUE_KEYS:
+                key_type = SINGLE_VALUE_KEY_TYPES[key]
+                fileH.write(f"{key} {type_value_map.get(key_type, key_type)}\n")
+                fileH.write(f"{key} {type_value_map.get(key_type, key_type)}\n")
```
Comment (andreas-el) [inline tests/unit_tests/c_wrappers/res/config/test_config_content.py:141]: This is probably obvious, but why do you write twice here? Is it to make sure keys are not duplicated at the assert step?

**Review by andreas-el — CHANGES_REQUESTED** [review]

File: tests/unit_tests/c_wrappers/res/config/test_config_content.py [hunk 7]
```diff
@@ -74,9 +112,35 @@ def test_config_content_as_dict(tmpdir):
...
-                str(Path.cwd() / existing_file_2),
-                str(Path.cwd() / existing_file_1),
-            ],
+            ConfigKeys.DATA_FILE: str(Path.cwd() / existing_file_2),
             ConfigKeys.REFCASE: str(Path.cwd() / existing_file_1),
         }
+
+
+def test_config_content_as_dict_single_value_keys(tmpdir):
+    with tmpdir.as_cwd():
```
Comment (valentin-krasontovitsch) [inline tests/unit_tests/c_wrappers/res/config/test_config_content.py:121]: sorry, quick question: i can't seem to figure out where this `tmpdir` is coming from - it seems like a fixture to me, but i cannot find its definition.

a short comment: the pattern of having a tmpdir and chdiring into it, to do stuff, is implemented in the `use_tmpdir` fixture and can be used by just adding `@pytest.mark.usefixtures("use_tmpdir")`
of course then you drop the argument `tmpdir`, I believe, and have to instead access the current directory by using `os.getcwd`... so as i'm not sure what `tmpdir` does exactly here, i'll leave it at your discretion - you probably know about `use_tmpdir` already anyways...

File: tests/unit_tests/c_wrappers/res/config/test_config_content.py (hunk 6 above)
Comment (valentin-krasontovitsch) [inline tests/unit_tests/c_wrappers/res/config/test_config_content.py:141]: the default value in the call to `get` is given in case an unexpected key (or rather a key with unexpected type) appears?

I'm sorry, maybe I'm a bit slow, but I don't quite understand parts of this test - one aspect is rather clear, the double entry that only yields a single value, as conveyed by the assert, but the necessity of the type value map is not immediately clear to me - is this purely technical, or does this also test something?

**Review by valentin-krasontovitsch — APPROVED** [review]
besides my little comments and questions, this looks good! it's nice to have a list of the single value keywords ^^

File: tests/unit_tests/c_wrappers/res/config/test_config_content.py (hunk 6 above)
Comment (DanSava) [reply tests/unit_tests/c_wrappers/res/config/test_config_content.py:141]: Yes, that is precisely why the key value map is written twice in the file.

File: src/ert/_c_wrappers/enkf/analysis_config.py (hunk 4 above)
Comment (DanSava) [reply src/ert/_c_wrappers/enkf/analysis_config.py:112]: > is it just tidying up, because we do not expect MAX_RUNTIME to be a list in the config dict?

Yes. 

Before the change, I was forced to add ln 104-106 and to guard against the possibility of users adding multiple instances of `MAX_RUNTIME` in the same config file. And "there can be only one" (movie quote) `MAX_RUNTIME` and now `config_content.as_dict()` makes sure this is the case.

File: tests/unit_tests/c_wrappers/res/config/test_config_content.py (hunk 7 above)
Comment (DanSava) [reply tests/unit_tests/c_wrappers/res/config/test_config_content.py:121]: `tmpdir` is a pytest default fixture. 

And it mainly does exactly what `@pytest.mark.usefixtures("use_tmpdir")` does only that you do not have to call `os.getcwd()` to get the directory you are in. 

Also, I am not a big fun of implementing functionality that I can get for free from a package we already require for tests.

Comment (valentin-krasontovitsch) [reply tests/unit_tests/c_wrappers/res/config/test_config_content.py:121]: that's why i couldn't find it ^^

ah, that makes a lot of sense! great, carry on : )

File: src/ert/_c_wrappers/enkf/analysis_config.py (hunk 4 above)
Comment (valentin-krasontovitsch) [reply src/ert/_c_wrappers/enkf/analysis_config.py:112]: ok cool, now i understand, thanks!

File: tests/unit_tests/c_wrappers/res/config/test_config_content.py (hunk 6 above)
Comment (valentin-krasontovitsch) [reply tests/unit_tests/c_wrappers/res/config/test_config_content.py]: OK i think i get it now - it's also a short way of supplying (valid according to type) values for all single value config keys. thanks for the explanation!

**Review by andreas-el — APPROVED** [review]

**Review by eivindjahren — APPROVED** [review]
:rocket:
